libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigation.java

```diff
@@ -611,6 +646,83 @@ public void removeRawLocationListener(@Nullable RawLocationListener rawLocationL
     navigationEventDispatcher.removeRawLocationListener(rawLocationListener);
   }
 
+  /**
+   * This adds a new enhanced location listener which is invoked when the best enhanced
+   * {@link android.location.Location} has been pushed. Either snapped (active guidance),
+   * map matched (free drive) or raw.
+   * <p>
+   * The behavior that causes this listeners callback to get invoked vary depending on whether
+   * free drive has been enabled using {@link #enableFreeDrive()} or disabled using
+   * {@link #disableFreeDrive()}.
+   * </p><p>
+   * It is not possible to add the same listener implementation more then once and a warning will be
+   * printed in the log if attempted.
+   * </p>
+   *
+   * @param enhancedLocationListener an implementation of {@code EnhancedLocationListener}
+   */
+  public void addEnhancedLocationListener(@NonNull EnhancedLocationListener enhancedLocationListener) {
+    navigationEventDispatcher.addEnhancedLocationListener(enhancedLocationListener);
+  }
+
+  /**
+   * This removes a specific enhanced location listener by passing in the instance of it or you can
+   * pass in null to remove all the listeners. When {@link #onDestroy()} is called, all listeners
+   * get removed automatically, removing the requirement for developers to manually handle this.
+   * <p>
+   * If the listener you are trying to remove does not exist in the list, a warning will be printed
+   * in the log.
+   *
+   * @param enhancedLocationListener an implementation of {@code EnhancedLocationListener}
+   */
+  public void removeEnhancedLocationListener(@Nullable EnhancedLocationListener enhancedLocationListener) {
+    navigationEventDispatcher.removeEnhancedLocationListener(enhancedLocationListener);
+  }
+
+  /**
+   * Calling this method enables free drive mode.
+   * <p>
+   * Best enhanced {@link Location} updates are received if an {@link EnhancedLocationListener} has been
+   * added using {@link #addEnhancedLocationListener(EnhancedLocationListener)}.
+   */
+  public void enableFreeDrive() {
+    isFreeDriveEnabled = true;
+    if (!isFreeDriveConfigured) {
+      freeDriveLocationUpdater.configure(applicationContext.getFilesDir(), "2019_04_13-00_00_11",
+          "https://api-routing-tiles-staging.tilestream.net", accessToken,
+          new OnOfflineTilesConfiguredCallback() {
+            @Override
+            public void onConfigured(int numberOfTiles) {
+              Timber.d("DEBUG: onConfigured %d", numberOfTiles);
+              isFreeDriveConfigured = true;
+              if (!isActiveGuidanceOnGoing) {
```

Comment:
We need to take into consideration that the configuration request is asynchronous. This means that in a current setup calling:
```
fun test() {
  enableFreeDrive()
  disableFreeDrive()
}
```
will still ultimately enable free drive and start requesting location updates when configuration is successful.

I believe that adding `isFreeDriveEnabled` check on this line should be sufficient if the calls and callbacks are all executed on the main thread. This should be verified with a test.